File: python/find_largest_value_in_each_tree_row_515.py
# Definition for a binary tree node.
from random import vonmisesvariate
from typing_extensions import Deque


class TreeNode:
 def __init__(self, val=0, left=None, right=None):
     self.val = val
     self.left = left
     self.right = right
class Solution:
     def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        out = []
        q = [root]
        visited = set()

        def bfs(node):
            if not node:
                return
            q = [node]
            visited.add(node)
            while q:
                next_q = []
                max_val = float('-inf')
                for node in q:
                    print(node.val)
                    max_val = max(max_val, node.val)
                    if node.left and node.left not in visited:
                        next_q.append(node.left)
                        visited.add(node.left)
                    if node.right and node.right not in visited:
                        next_q.append(node.right)
                        visited.add(node.right)
                out.append(max_val)
                q = next_q

            bfs(root)
            

File: rust/src/find_largest_value_in_each_tree_row_515.rs
use crate::Solution;
use crate::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn largest_values(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut res = Vec::new();
        Self::bfs(root, &mut res);
        res
    }

    fn bfs(root: Option<Rc<RefCell<TreeNode>>>, res: &mut Vec<i32>) {
        if root.is_none() {
            return;
        }

        let mut q = vec![root];
        while !q.is_empty() {
            let mut next_q = vec![];
            let mut max_val = i32::MIN;

            for node_opt in q.iter() {
                if let Some(node_rc) = node_opt {
                    let node = node_rc.borrow();
                    max_val = max_val.max(node.val);

                    if let Some(left_rc) = node.left.clone() {
                        next_q.push(Some(left_rc));
                    }
                    if let Some(right_rc) = node.right.clone() {
                        next_q.push(Some(right_rc));
                    }
                }
            }

            res.push(max_val);
            q = next_q;
        }
    }
}
